src/cascadia/TerminalApp/CommandPalette.cpp

```diff
@@ -343,7 +344,56 @@ namespace winrt::TerminalApp::implementation
         {
             auto top = heap.top();
             heap.pop();
-            _filteredActions.Append(top.command);
+            actions.push_back(top.command);
+        }
+
+        return actions;
+    }
+
+    // Method Description:
+    // - Update our list of filtered actions to reflect the current contents of
+    //   the input box. For more details on which commands will be displayed,
+    //   see `_getWeight`.
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - <none>
+    void CommandPalette::_updateFilteredActions()
+    {
+        auto actions = _collectFilteredActions();
+
+        // Make _filteredActions look identical to actions, using only Insert and Remove.
+        // This allows WinUI to nicely animate the ListView as it changes.
+        for (uint32_t i = 0; i < _filteredActions.Size() && i < actions.size(); i++)
+        {
+            for (uint32_t j = i; j < _filteredActions.Size(); j++)
+            {
+                if (_filteredActions.GetAt(j) == actions[i])
+                {
+                    for (uint32_t k = i; k < j; k++)
+                    {
+                        _filteredActions.RemoveAt(i);
+                    }
+                    break;
+                }
+            }
+
+            if (_filteredActions.GetAt(i) != actions[i])
+            {
+                _filteredActions.InsertAt(i, actions[i]);
+            }
+        }
+
+        // Remove any extra trailing items from the destination
+        while (_filteredActions.Size() > actions.size())
+        {
+            _filteredActions.RemoveAtEnd();
+        }
+
+        // Add any extra trailing items from the source
+        while (_filteredActions.Size() < actions.size())
+        {
+            _filteredActions.Append(actions[_filteredActions.Size()]);
```

Comment:
I trust that this all works because you tested it, but I just can't for the life of me figure out how and why iterating over in index order doesn't explode the moment we start manipulating the array layout underneath us o_O

Reply:
Okay, this is a process by which we successively close gaps where `_filteredActions` doesn't match `actions` and insert `actions` that aren't in the right place in `_filteredActions`. After that, we just bulk remove `_filteredActions` that are off the end (because the real actions pushed them out) and bulk insert the `actions` that weren't represented.

Fascinating!

Reply:
Yeah sorry, I know it's opaque, and it has a "rest of the owl" moment in there.  I wasn't quite sure what to put in the comments to make it clear how it works, without drowning the code in a wall of text.

![image](https://user-images.githubusercontent.com/10259764/87742068-d7584a00-c79a-11ea-9b22-ff3f76aa1cdc.png)